Support for u128/i128

I would like to convert values which may be 128bit integers, which currently unsupported.

As I see you use __cast__ crate for conversion, which looks like unmaintained.

Cast have optional support for 128bit integers which activates by _x128_ feature, but it doesn't work with stable rust due to:

```
error[E0554]: #![feature] may not be used on the stable release channel
  --> /home/kayo/.cargo/registry/src/github.com-1ecc6299db9ec823/cast-0.2.2/src/lib.rs:99:31
   |
99 | #![cfg_attr(feature = "x128", feature(i128_type, i128))]
   |                               ^^^^^^^^^^^^^^^^^^^^^^^^
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -12,7 +12,10 @@ serde = "1.*"
 error-chain = "0.12.*"
 neon = "0.3.1"
 neon-runtime = "0.3.1"
-cast = "0.2.*"
+
+[dependencies.num]
+version = "0.2"
+default-features = false
 
 [dev-dependencies]
 serde_derive = "1.*"

--- a/src/de.rs
+++ b/src/de.rs
@@ -152,7 +152,7 @@ impl<'x, 'd, 'a, 'j, C: Context<'j>> serde::de::Deserializer<'x> for &'d mut Des
 
     forward_to_deserialize_any! {
        <V: Visitor<'x>>
-        bool i8 i16 i32 i64 u8 u16 u32 u64 f32 f64 char str string
+        bool i8 i16 i32 i64 i128 u8 u16 u32 u64 u128 f32 f64 char str string
         unit unit_struct seq tuple tuple_struct map struct identifier
         newtype_struct
     }

--- a/src/errors.rs
+++ b/src/errors.rs
@@ -65,11 +65,11 @@ error_chain! {
             description("JS exception")
             display("JS exception")
         }
-    }
-
-    foreign_links {
-        NumberCastError(::cast::Error)
-        #[doc = r"occurs when deserializing a number would cause an over/under flow"];
+        // failed to convert something to f64
+        CastError {
+            description("CastError")
+            display("CastError")
+        }
     }
 }
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -70,10 +70,10 @@
 //! ```
 //!
 
-extern crate cast;
 #[macro_use]
 extern crate error_chain;
 extern crate neon;
+extern crate num;
 #[macro_use]
 extern crate serde;
 

--- a/src/ser.rs
+++ b/src/ser.rs
@@ -135,6 +135,12 @@ where
     }
 
     #[inline]
+    fn serialize_i128(self, v: i128) -> Result<Self::Ok, Self::Error> {
+        Ok(JsNumber::new(self.cx, as_num::<_, f64>(v)?).upcast())
+    }
+
+
+    #[inline]
     fn serialize_u8(self, v: u8) -> Result<Self::Ok, Self::Error> {
         Ok(JsNumber::new(self.cx, as_num::<_, f64>(v)?).upcast())
     }
@@ -151,6 +157,11 @@ where
 
     #[inline]
     fn serialize_u64(self, v: u64) -> Result<Self::Ok, Self::Error> {
+        Ok(JsNumber::new(self.cx, as_num::<_, f64>(v)?).upcast())
+    }
+
+    #[inline]
+    fn serialize_u128(self, v: u128) -> Result<Self::Ok, Self::Error> {
         Ok(JsNumber::new(self.cx, as_num::<_, f64>(v)?).upcast())
     }
 

--- a/src/ser.rs
+++ b/src/ser.rs
@@ -2,13 +2,20 @@
 //! Serialize a Rust data structure into a `JsValue`
 //!
 
-use cast;
 use errors::Error;
 use errors::ErrorKind;
 use errors::Result as LibResult;
 use neon::prelude::*;
 use serde::ser::{self, Serialize};
 use std::marker::PhantomData;
+use num;
+
+fn as_num<T: num::cast::NumCast, OutT: num::cast::NumCast>(n: T) -> LibResult<OutT> {
+    match num::cast::<T, OutT>(n) {
+        Some(n2) => Ok(n2),
+        None => bail!(ErrorKind::CastError)
+    }
+}
 
 /// Converts a value of type `V` to a `JsValue`
 ///
@@ -109,47 +116,47 @@ where
 
     #[inline]
     fn serialize_i8(self, v: i8) -> Result<Self::Ok, Self::Error> {
-        Ok(JsNumber::new(self.cx, cast::f64(v)).upcast())
+        Ok(JsNumber::new(self.cx, as_num::<_, f64>(v)?).upcast())
     }
 
     #[inline]
     fn serialize_i16(self, v: i16) -> Result<Self::Ok, Self::Error> {
-        Ok(JsNumber::new(self.cx, cast::f64(v)).upcast())
+        Ok(JsNumber::new(self.cx, as_num::<_, f64>(v)?).upcast())
     }
 
     #[inline]
     fn serialize_i32(self, v: i32) -> Result<Self::Ok, Self::Error> {
-        Ok(JsNumber::new(self.cx, cast::f64(v)).upcast())
+        Ok(JsNumber::new(self.cx, as_num::<_, f64>(v)?).upcast())
     }
 
     #[inline]
     fn serialize_i64(self, v: i64) -> Result<Self::Ok, Self::Error> {
-        Ok(JsNumber::new(self.cx, cast::f64(v)).upcast())
+        Ok(JsNumber::new(self.cx, as_num::<_, f64>(v)?).upcast())
     }
 
     #[inline]
     fn serialize_u8(self, v: u8) -> Result<Self::Ok, Self::Error> {
-        Ok(JsNumber::new(self.cx, cast::f64(v)).upcast())
+        Ok(JsNumber::new(self.cx, as_num::<_, f64>(v)?).upcast())
     }
 
     #[inline]
     fn serialize_u16(self, v: u16) -> Result<Self::Ok, Self::Error> {
-        Ok(JsNumber::new(self.cx, cast::f64(v)).upcast())
+        Ok(JsNumber::new(self.cx, as_num::<_, f64>(v)?).upcast())
     }
 
     #[inline]
     fn serialize_u32(self, v: u32) -> Result<Self::Ok, Self::Error> {
-        Ok(JsNumber::new(self.cx, cast::f64(v)).upcast())
+        Ok(JsNumber::new(self.cx, as_num::<_, f64>(v)?).upcast())
     }
 
     #[inline]
     fn serialize_u64(self, v: u64) -> Result<Self::Ok, Self::Error> {
-        Ok(JsNumber::new(self.cx, cast::f64(v)).upcast())
+        Ok(JsNumber::new(self.cx, as_num::<_, f64>(v)?).upcast())
     }
 
     #[inline]
     fn serialize_f32(self, v: f32) -> Result<Self::Ok, Self::Error> {
-        Ok(JsNumber::new(self.cx, cast::f64(v)).upcast())
+        Ok(JsNumber::new(self.cx, as_num::<_, f64>(v)?).upcast())
     }
 
     #[inline]
@@ -174,7 +181,7 @@ where
 
     #[inline]
     fn serialize_bytes(self, v: &[u8]) -> Result<Self::Ok, Self::Error> {
-        let mut buff = JsBuffer::new(self.cx, cast::u32(v.len())?)?;
+        let mut buff = JsBuffer::new(self.cx, as_num::<_, u32>(v.len())?)?;
         self.cx.borrow_mut(&mut buff, |buff| buff.as_mut_slice().clone_from_slice(v));
         Ok(buff.upcast())
     }

--- a/test/__tests__/get_values.js
+++ b/test/__tests__/get_values.js
@@ -50,7 +50,9 @@ describe('all values ok', () => {
             l: "jkl",
             m: [0,1,2,3,4],
             o: {Value: ['z', 'y', 'x']},
-            p: [1, 2, 3.5]
+            p: [1, 2, 3.5],
+            q: 999,
+            r: 333,
         });
     });
 
@@ -79,7 +81,9 @@ describe('all values ok', () => {
             l: "jkl",
             m: [0,1,2,3,4],
             o: {Value: ['z', 'y', 'x']},
-            p: [1, 2, 3.5]
+            p: [1, 2, 3.5],
+            q: 999,
+            r: 333,
         };
 
         o.self = o;
@@ -123,7 +127,9 @@ describe('all values ok', () => {
             l: "jkl",
             m: [0,1,2,3,4],
             o: {Value: ['z', 'y', 'x']},
-            p: [1, 2, 3.5]
+            p: [1, 2, 3.5],
+            q: 999,
+            r: 333,
         };
         const o2 = native.roundtrip_object(o);
         expect(o).toEqual(o2);

--- a/test/native/src/lib.rs
+++ b/test/native/src/lib.rs
@@ -45,6 +45,8 @@ struct AnObjectTwo {
     m: Vec<u8>,
     o: TypeEnum,
     p: Vec<f64>,
+    q: u128,
+    r: i128,
 }
 
 macro_rules! make_test {
@@ -103,6 +105,8 @@ make_test!(make_object, {
         m: vec![0, 1, 2, 3, 4],
         o: TypeEnum::Value(vec!['z', 'y', 'x']),
         p: vec![1., 2., 3.5],
+        q: 999,
+        r: 333
     };
     value
 });
@@ -150,6 +154,8 @@ make_expect!(
         m: vec![0, 1, 2, 3, 4],
         o: TypeEnum::Value(vec!['z', 'y', 'x']),
         p: vec![1., 2., 3.5],
+        q: 999,
+        r: 333
     },
     AnObjectTwo
 );

